### PR TITLE
 API Platform support

### PR DESCRIPTION
**Pull request by SirDomin**
By using plugin class we can access its data by request from ApiPlatform within Sylius api platform payment integration.
Some refactor is needed inside PayPalPlugin.

### CODE REVIEW

File: src/ApiPlatform/PayPalPayment.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            'completePayPalOrderFromPaymentPageUrl' => $this->router->generate(
+                    'sylius_paypal_plugin_complete_paypal_order',
+                    ['token' => $order->getTokenValue()],
+                    UrlGeneratorInterface::ABSOLUTE_URL
+                ),
+            'createPayPalOrderFromPaymentPageUrl' => $this->router->generate(
+                    'sylius_paypal_plugin_create_paypal_order',
+                    ['token' => $order->getTokenValue()],
+                    UrlGeneratorInterface::ABSOLUTE_URL
+                ),
```
Comment (Zales0123) [inline src/ApiPlatform/PayPalPayment.php:70]: Here as well

File: src/ApiPlatform/PayPalPayment.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+use Sylius\Bundle\PayumBundle\Model\GatewayConfigInterface;
+use Sylius\Component\Core\Model\OrderInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+use Sylius\PayPalPlugin\Provider\AvailableCountriesProviderInterface;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Routing\RouterInterface;
+
+//TODO: once we implement ApiPlatformPayment interface in core, implement it here
```
[the remark this answers is not in the document]
Comment (Zales0123) [reply src/ApiPlatform/PayPalPayment.php:24]: And I would either wait for it or (better) implement a custom interface here and then extend the mentioned one in it

File: src/ApiPlatform/PayPalPayment.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+final class PayPalPayment
```
Comment (Zales0123) [inline src/ApiPlatform/PayPalPayment.php:25]: We're also missing some spec or test for this class

File: src/Resources/config/services.xml [hunk 4]
```diff
@@ -26,6 +26,12 @@
             class="Sylius\PayPalPlugin\Generator\PayPalAuthAssertionGenerator"
         />
 
+        <service id="Sylius\PayPalPlugin\ApiPlatform\PayPalPayment">
+            <argument type="service" id="router" />
+            <argument type="service" id="Sylius\PayPalPlugin\Provider\AvailableCountriesProviderInterface" />
+            <tag name="api.payment" />
```
Comment (Zales0123) [inline src/Resources/config/services.xml:32]: Can you explain what does this tag do?

**Review by Zales0123 — COMMENTED** [review]
I feel a little bit worried that we're introducing something that has no use in the plugin itself :( Can we at least test it with some integration/functional test?

File: src/Resources/config/services.xml (hunk 4 above)
Comment (SirDomin) [reply src/Resources/config/services.xml:32]: this tag is used to collect all data providers from payment plugins, in Core we got PaymentConfiguration provider which allows us to send data to frontend based on picked payment method, f.eg if we pick paypal payment method we search throughout all payment methods with api.payment tag and see which one supports it, then we send payment data configuration to pay using specific payment method.

File: src/ApiPlatform/PayPalPayment.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (SirDomin) [reply src/ApiPlatform/PayPalPayment.php:24]: there will be implemented custom interface which will allow us to remove api.payment tag from services in PayPalPlugin, it all will be loaded automatically once we implement this interface in Core

[the remark this answers is not in the document]
Comment (SirDomin) [reply src/ApiPlatform/PayPalPayment.php:24]: https://github.com/Sylius/Sylius/pull/12107

**Review by Zales0123 — APPROVED** [review]

**Comment by Zales0123** [comment]
Thanks, @SirDomin! :1st_place_medal:

**Pull request merged**